### ISSUE TITLE
[Feature request] Allow TinyMCE to fill container - 100% height

### ISSUE THREAD

**Issue by Misiu** [tinymce/tinymce#3546]
**Do you want to request a *feature* or report a *bug*?**
feature
**What is the current behavior?**
currently we can specify height of TinyMCE as number, but there isn't any easy way to set height to `100%`
**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem via [fiddle.tinymce.com](http://fiddle.tinymce.com/) or similar.

**What is the expected behavior?**
I'd like to be able to specify that TinyMCE should fill it's parent container. If it is inside div then it should fill div, if inside body then entire body.
**Which versions of TinyMCE, and which browser / OS are affected by this issue? Did this work in previous versions of TinyMCE?**
4.5.5 Chrome 57 Win10x64

There are many questions related to this issue, for example
https://community.tinymce.com/communityQuestion?id=90661000000MrXkAAK
http://stackoverflow.com/questions/9588025/change-tinymce-editors-height-dynamically
http://stackoverflow.com/questions/19861838/tinymce-4-and-100-height-within-containing-element#

and some partly solutions:
https://gist.github.com/xypaul/efae27e0d0093d0746c3

but they aren't working perfectly, In my case my custom sidebar isn't displayed correctly.

It would be awesome if You guys could allow this behavior, simply by adding an extra option (fill_container, default false) for it or allow us to specify height as percentage value.

When using resize we can resize editor, so maybe same code can be used, but instead of user selected height it can use height of parent container?

**Comment by Misiu** [comment]
I've created POC: http://codepen.io/Misiu/pen/EWbpyr, it is based on [this SO answer](http://stackoverflow.com/a/25208334/965722).
I've checked this in Chrome 57, IE 11 and EDGE and it works fine.
Hopefully similar functionality can be added to core.

**Comment by Misiu** [comment]
I've created plugin for this:

```javascript
tinymce.PluginManager.add('autoheight', function(editor) {

  /**
   * Check if plugin is enabled
   * @type {bool}
   */
  var autoheight_enabled = editor.getParam('auto_height', false);
  var editor_container;
  var bars_height = 0;

  function resizeEditor(e) {
    if (typeof(editor_container) === 'undefined') return;

    try {
      var element_height = parseInt(window.getComputedStyle(editor_container).height);
      /*calculate bar height only once*/
      if (bars_height === 0) {
        var toolbars = document.querySelectorAll('.mce-toolbar, .mce-statusbar, .mce-menubar');
        toolbars.forEach(function(toolbar, i) {
          /*skip sidebar*/
          if (!toolbar.classList.contains('mce-sidebar-toolbar')) {
            var bar_height = parseInt(window.getComputedStyle(toolbar).height);
            bars_height += bar_height;
          }
        });
      }
      /*the extra 8 is for margin added between the toolbars*/
      new_height = element_height - bars_height - 8; 
      editor.theme.resizeTo('100%', new_height);
    } catch (err) {
      console.log(err);
    }

  }

  function handleResizeWindow(e) {
    if (autoheight_enabled) {
      resizeEditor(e);
    }
  }

  editor.on('ResizeWindow', handleResizeWindow);
  editor.on("init", function() {
    try {
      editor_container = editor.getContainer().parentNode;
    } catch (e) {

    }
    if (autoheight_enabled) {
      setTimeout(function() {
        resizeEditor();
      }, 0);
    }
  });
});
```

Demo is available here: https://codepen.io/Misiu/pen/EWEVPo

I'll create repository for this if this can't be added to core.

**Comment by Misiu** [comment]
Any chance this might be added to core? Or as official plugin?

**Comment by donShakespeare** [comment]
@Misiu Is this related?
https://github.com/tinymce/tinymce/issues/3635

**Comment by Misiu** [comment]
@donShakespeare not quite. Issue You referred is about **width**, what I want is **height**.
My idea is to have a div (container) and have tinyMCE fill that div, so when I resize that div tinyMCE will resize.
As I mentioned I've created a plugin for this, but I think this could be added to core.

**Comment by kiesslir** [comment]
I have the same problem, but in my case `window.getComputedStyle(editor_container).height` returns `auto` and therefore `new_height` is **NaN**

**Comment by Misiu** [comment]
@kiesslir have You tried my demo? https://codepen.io/Misiu/pen/vZEGqm
Please provide sample that shows Your behavior. Maybe I can help by updating my plugin.
It would be awesome if TinyMCE would have this build in.

**Comment by donShakespeare** [comment]
@Misiu great plugin

**Comment by Misiu** [comment]
@donShakespeare thanks 😄 this is just a prototype, but for me it works fine. All improvements are more than welcome.

**Comment by kiesslir** [comment]
@Misiu I use tinyMCE inside a jQuery dialog. On page load, the editor area is not initialised. I modified your version so that's working for me.

**stand alone example**
[tinyMCE.html.txt](https://github.com/tinymce/tinymce/files/1055619/tinyMCE.html.txt)

**Comment by sixtyfive** [comment]
Awesome, works like a charm. Thank you @Misiu!

**Comment by Mihaiii** [comment]
@Misiu That's a great plugin, thanks! My only question is why do you "calculate bar height only once"? If you shrink the window and have multiple buttons in the toolbar, the height of the toolbar changes.

**Comment by Misiu** [comment]
@Mihaiii thank You for finding this. I build this plugin as a proof of concept. Ideally this would be build in into TinyMCE. 
I think I'll create repo for this plugin, so it will be easier to fix such bugs.

**Comment by sixtyfive** [comment]
@Mihaiii is there any chance of this feature becoming part of TinyMCE as @Misiu suggests (and surely others agree)?

**Comment by Mihaiii** [comment]
@sixtyfive I think this was a misunderstanding. I don't maintain TinyMCE :).

**Comment by Misiu** [comment]
@fyrkant maybe You could take a look at this?

**Comment by Misiu** [comment]
@fyrkant any updates?
I think TinyMCE is awesome, but without proper fill parent solution it isn't perfect.
In my case im using TinyMCE in split panel container.
I'm calculating everything by hand, but maybe this can be solved with pure CSS solution? This would be even better and would remove glitching (when I resize panel fast status bar in TinyMCE isn't always fixed to bottom, it needs second to update it's position)

**Comment by sixtyfive** [comment]
@Misiu no misunderstanding, I was just hoping that as the author of the plugin you might have some info on whether it was received at all by the TinyMCE people :)

**Comment by Misiu** [comment]
@sixtyfive I know as much as You :) I've only noticed feature label, nothing more.
It's sad that such easy features with working code aren't added.
I've tried couple of times getting TinyMCE people attention on this but without any luck.

**Comment by nileshbhagwat** [comment]
In "Inline" mode, editor.getContainer() is always null.

tinymce.dom.DomQuery('.mce-toolbar', editor.getContainer());
so for this query it return toolbars from all editors, (since editor.getContainer() is null)

Is there any way to get editor specific toolbars for inline mode?

**Comment by BES76** [comment]
I'm late to the party... no need for a plugin... couple of lines in CSS will fix the sizing problem. The basic idea is to transform the plugin divs into a table.

div.mce-container { manipulate hight/width here }
div.mce-container > div { display: table; height: 100%; width: 100%; }
div.mce-container > div > div { display: table-row; }
div.mce-container > div > div:nth-of-type(1) > div { display: table-cell; border: 1px solid #D5D5D5; background: #f0f0f0; }
div.mce-container > div > div:nth-of-type(2) { height: 100%; border: none; }
div.mce-container > div > div:nth-of-type(2) iframe { display: table-cell; width: 100% !important; height: 100% !important; border-left: 1px solid #D5D5D5; border-right: 1px solid #D5D5D5; box-sizing: border-box; }
div.mce-container > div > div:nth-of-type(3) > div:nth-of-type(1) { display: table-cell; border: 1px solid #D5D5D5; background: #f0f0f0; }

**Comment by Misiu** [comment]
@BES76 can You share a demo showing this working? I can't get that to work.

**Comment by Afraithe** [comment]
@Misiu did you create a pull request for this plugin?

Now that we have dropped some old browser support, this might be interesting to pull in.

**Comment by BES76** [comment]
@Misiu 
Sorry, wrong class it should be "mce-tinymce" not "mce-container"
You can check it out here: [https://knowhow.bwise.ch/tiny/](url)

**Comment by Misiu** [comment]
@Afraithe unfortunately not. I'm not that advanced JS developer.
My latest version is available at https://codepen.io/Misiu/pen/vZEGqm.

As @nileshbhagwat mentioned this might not work for inline mode.
Also @kiesslir added some changes so that in works in dialog.

@Afraithe You can reuse my code if You want, or I can create PR for this, but I won't be able to provide tests (I didn't wrote any for JS)

**Comment by Misiu** [comment]
@BES76 thanks for update.
@Afraithe maybe CSS only solution would be better? This would eliminate size calculation in JS.
What do You think? This way my plugin would add css class only and rest would be done by CSS and browser.

**Comment by Afraithe** [comment]
Yes perhaps, it might be better to do it in modern css now days, but we still probably need to implement an option for it in JS, then add classes to support it.

We are doing quite extensive work on the UI framework over the next couple of months, can't guarantee we can work this in right away (deadline already tight) but it would be nice to have for sure.

**Comment by Misiu** [comment]
@Afraithe ideally `autoheight` plugin should only add class to main TinyMCE container, rest should by done by CSS.
I can try hacking something using code by @BES76.
I'll share results when something will be working :)

**Comment by BES76** [comment]
@Misiu don't forget that table-cell elements can't have position relative, so you need a different way to position the "powered by" label. 
.mce-branding-powered-by { top: auto; right 2px; bottom 38px; }
Unless you are like me and like it at the top.

**Comment by Misiu** [comment]
@Afraithe anything changed in TinyMCE 5 regarding 100% height?

**Comment by Misiu** [comment]
@Afraithe any updates? 🙂

**Comment by spocke** [comment]
Here is a somewhat hacky demo of doing this: http://fiddle.tinymce.com/fIgaab will see if we can make official support for this it looks like the only thing we need to add is a flex: 1 on the parent container.

**Comment by Misiu** [comment]
@spocke I've tweaked Your demo by changing width and height to 100% (http://fiddle.tinymce.com/fIgaab/2) but unfortunately it doesn't work for me.

**Comment by spocke** [comment]
You can use flex on the body and position it to fill the whole area http://fiddle.tinymce.com/fIgaab/3

**Comment by spocke** [comment]
Positioning the body is generally a bad idea this is much cleaner: http://fiddle.tinymce.com/fIgaab/4

**Comment by Misiu** [comment]
@spocke thats much better :)
As You mentioned it would be awesome to have official support - to be able to specify `100%` as width and height.

**Comment by Misiu** [comment]
@spocke any updates? :)

**Comment by IonelLupu** [comment]
Is this still opened? Why?

**Comment by maifeeulasad** [comment]
For those, who are still struggling with it:
```
.tox-tinymce{
  height: 100% !important;
}
```
*I tested it on a react project.*

**Comment by Misiu** [comment]
@Afraithe ping :)

**Comment by lostkeys** [comment]
@Misiu, doesn't simply setting the height config option to 100% solve your issue? Here is an example: https://fiddle.tiny.cloud/uRhaab

**Comment by Misiu** [comment]
@lostkeys I must admit I checked that a long time ago, but in the sample you linked it works.
I only had to hide resize handler using:
```
[title="Resize"],
.tox-statusbar__resize-handle {
    display: none !important;
}
```

I'll check the newest version in my scenario (standard usage, plus in popups) and will report back if I have problems.

**Comment by lnewson** [comment]
@Misiu as an extra tip you can disable the resize handle by adding the `resize: false` configuration which is our supported way to do that. For example: https://fiddle.tiny.cloud/uRhaab/1

**Comment by AlexanderKugler** [comment]
Hi,
Just wanted to politely follow up on this issue and ask whether there is now an official, documented solution for making the editor fill its parent container (100% height), or whether this is still considered unresolved.
